rfcs/0164-bootspecv2.md

```diff
@@ -0,0 +1,300 @@
+)) and
+issues with booting in the Heads platform (e.g., [Issue #1348](https://github.com/linuxboot/heads/issues/1348.
+)).
+
+Initrd secrets serve to protect boot-time secrets from exposure within the Nix
+store. This is achieved by using runtime scripts that append CPIO archives to
+the generation's initrd during bootloader installation. However, the term
+"initrd secrets" can be misleading, as the secrets are plaintext and can be
+accessed from the ESP partition, offering limited confidentiality and no
+integrity.
+
+In practice, initrd secrets are often employed to establish stable fingerprints
+for SSH servers within the initrd, aiding in remote disk decryption on servers.
+To address these issues, this RFC proposes moving away from the "appender
+script" model in Bootspec v1 and instead adopting a hash map format to
+represent secrets and their corresponding values.
+
+This change allows for greater flexibility and enhanced security. For users in
+the NixOS ecosystem relying on systemd and its semantics, the proposal suggests
+offering a systemd-credentials approach for handling initrd secrets,
+potentially enabling encryption using TPM2 if available, or a straightforward
+key if TPM2 is not present. Additionally, this approach provides the means to
+build using the `LoadEncryptedCredential` abstraction.
+
+For those not using systemd, the hash map format offers full flexibility to
+bootloader implementations to determine how to add the secrets, such as
+appending to the initrd or other methods.
+
+**Key Takeaway**: This RFC recommends changing the format of Bootspec v2 to support
+`initrdSecrets` as a hashmap with string keys and values. The key represents a
+name used to organize secrets, and the value denotes an accessible path during
+bootloader installation.
+
+_Example_
+
+```
+"initrdSecrets": {
+   "my-private-key": "/etc/nixos/secrets/wireguard-key"
+}
+```
+
+## Device Tree and Device Tree Overlays
+
+Non-x86 systems often rely on device trees to inform firmware or bootloader about the available devices and hardware support. 
+
+There are two distinct requirements for device trees: 
+
+- generic hardware support, where firmware or bootloader selects the appropriate device tree
+- device-specific support, where a hardcoded device tree is required. 
+
+The latter indicates a potentially problematic, non-upstreamed, or in-development platform.
+
+The current Bootspec v1 does not formally encode information about hardcoded device trees or the folder containing available device trees. As a result, unformalized extensions are needed to address these fundamental use cases.
+
+Regarding overlays in NixOS, any expressed overlay is incorporated into the final device tree. This eliminates the necessity to formalize an additional overlays field list, as overlays can be transformed into device trees as needed.
```

Comment:
See earlier you complained about devicetrees being hardcoded, and then here you say "btw we think non-hardcoded devicetrees are lame so we won't allow them".  It's circular reasoning...

If you want to keep things simple and require one single `.dtb` file for bootspec v2 that's fine, since a single `dtb` is better than none at all.  Just please don't portray limitations of the simplifying assumption you're making as being limitations of devicetrees.

At the end of the day devicetrees are basically a cleaner version of ACPI without the halting problem or marketing budget.  They're data, not code.  They don't have bigtech employees running around trumpeting their virtues.  All the other differences are basically just syntax.

Reply:
Well, foremost, DTOs are exactly implemented like this in NixOS.

I don't think they are lame, but it has been brought to me again by @samueldr that DTOs are not necessarily the panacea we have been looking for and it's unclear to me what would be the provided benefits of supporting a first-class field for DTOs inside Bootspec.

So please enlighten me:

> If you want to keep things simple and require one single .dtb file for bootspec v2 that's fine, since a single dtb is better than none at all. Just please don't portray limitations of the simplifying assumption you're making as being limitations of devicetrees.

Can you mention some examples of introduced limitations via our simplifying assumptions we are doing by bolting any DTOs inside DTBs ?

I had some ideas of my own: multiple-devices variants support, etc. But I am not convinced they are not supported with the current proposal.

Also, please get me right: I'm not saying they are lame, I'm saying we can re-encode any DTOs as a DTB, therefore, we don't need (until proven wrong?) DTOs as part of the generation data, but we should "erase" it via the NixOS module machinery.

Reply:
Is there more than whining and FUD that you want to apply to this section, @amjoseph-nixpkgs?

Reminder that bootspec is about *listing the knowledge from a NixOS generation for configuring boot*.

There are no generic methods of describing overlays that will apply to generic boot methods.

Or maybe you can elucidate this whole situation and show what is the generic device tree overlay spec in practice, and what form of ***knowledge*** that is needed to expose in bootspec to *allow configuring the boot for a specific given generation*.